recbole/trainer/trainer.py

```diff
@@ -134,6 +152,43 @@ def _build_optimizer(self, params):
             optimizer = optim.Adam(params, lr=self.learning_rate)
         return optimizer
 
+    ###@Juyong Jiang
+    def _build_distribute(self, backend):
+        # 1 set backend
+        torch.distributed.init_process_group(backend=backend)
+        # 2 get distributed id
+        local_rank = torch.distributed.get_rank()
+        torch.cuda.set_device(local_rank)
+        device_dis = torch.device("cuda", local_rank)
+        # 3, 4 assign model to be distributed
+        self.model.to(device_dis)
+        self.model = DistributedDataParallel(self.model, 
+                                             device_ids=[local_rank],
+                                             output_device=local_rank).module
+        return self.model
+
+    def _trans_dataload(self, interaction):
+        data_dict = {}
+        #using pytorch dataload to re-wrap dataset
+        def sub_trans(dataset):
+            dis_loader = DataLoader(dataset=dataset,
+                                    batch_size=dataset.shape[0],
+                                    sampler=DistributedSampler(dataset, shuffle=False))
+            for data in dis_loader:
+                batch_data = data
```

Comment:
I do not understand this loop, it seems `batch_data` will be the last 'data' of `dis_loader`, could you please explain it?

Reply:
And did you test your code in some datasets like ml-100k? Could you provide us the performance results of models? I want to know if the model performance will change a lot compared with single-GPU training.

Reply:
Hi, Xingyu! Yeah, my pleasure! In our `DataLoader` class, I assign `batch_size=dataset.shape[0]` that means it extracts all data in current `batch_size`.  So the length of `dis_loader` will be only one, i.e. like this  `for data in range(1)`.

https://github.com/juyongjiang/RecBole/blob/0d35771629f65a9a06ad7e66dd11bfbe06091971/recbole/trainer/trainer.py#L173-L180

Reply:
Yeah, of course! Please wait for a moment! I will provide a table to illustrate the performance compared with single GPU training.

Reply:
@2017pxy Hi, Xingyu! I have got the experimental results. It seems that it doesn't decrease the performance a lot but significantly reduces the training time by **about 3.78 times**. BTW, I just run the experiment only one time. So I think this performance drift can be ignored. : )
![image](https://user-images.githubusercontent.com/50765248/132860103-a5e874b5-7c13-44b4-a2ed-98a120808917.png)

Reply:
@2017pxy Any further questions and or comments? Thanks in advance.

Reply:
Hi @juyongjiang @hunkim, sorry for late reply.

Following your implementation, our team modified the `trainer` and made some tests. We find your implementation works well for model training. Thanks for your contribution!

However,  since the time cost of `run_recbole` is mainly from model evaluation, we want to implement the multi-GPUs  evaluation as well and release together with the multi-GPUs training. Unfortunately,  we face some problems when we apply your implementations to evaluation since the data organization for evaluation is different. Thus,  I am sorry to tell you that it still takes some time to release this new feature, and even this new feature might not be added in next version.

Thanks again for your implementation, and if you have any idea and  suggestions about multi-GPUs  evaluation, please tell us know.

Reply:
@2017pxy Okay, got it! Thanks for your reply. I will implement the multi-GPUs evaluation as well and pull a new request. : )